File: 04_Non-ComparisonSorts/Radix Sort/RadixSort.cpp
#include <iostream>
#include <vector>
#include <time.h>
#include <algorithm>
using namespace std;

vector<int> initializeArrayWithRandomNumbers(int n)
{
    srand (time(NULL));
    vector<int> vec;
    vec.push_back(0);
    for(int i = 1; i < n+1; ++i)
    {
        vec.push_back(rand()%n+1);
    }

    return vec;
}

void swap(int& val1, int& val2)
{
    int temp = val1;
    val1 = val2;
    val2 = temp;
}

void printVector(vector<int>vec)
{
    for(int i = 1; i<vec.size(); ++i)
    {
        cout<<vec.at(i)<<"\t";
    }

    cout<<endl;
}

int findMax(vector<int> a)
{
    int max = a.at(0);
    for(int i = 1; i < a.size(); ++i)
    {
        if (max < a.at(i))
        max = a.at(i);
    }
    return max;
}

void countingSort(vector<int>&invec, int exp, vector<int>&outvec)
{
	vector<int>cumulativeSum;
    int max = findMax(invec);// find the max element in the input vector
    cumulativeSum.resize(max+1);
    outvec.resize(invec.size());

    for(int i = 0; i <= max; ++i)
    {
        cumulativeSum.at(i) = 0;
    }

    for(int i = 1; i < invec.size(); ++i)
    {
        ++ cumulativeSum.at(invec.at(i));
    }

    for(int i = 1; i <= max; ++i)
    {
        cumulativeSum.at(i) += cumulativeSum.at(i-1);
    }

    for(int i = invec.size() - 1; i >= 1; --i)
    {
        outvec.at(cumulativeSum.at(invec.at(i))) = invec.at(i);
        -- cumulativeSum.at(invec.at(i));
    }

}

void radixSort(vector<int>&invec, vector<int>&outVec)
{
	int m = findMax(invec);
	for (int exp = 1; m / exp > 0; exp *= 10)
        countingSort(invec, exp, outVec);
}

int main()
{
    cout<<"The Array to be Sorted\n";
    int size = 10;
    vector<int>vec = initializeArrayWithRandomNumbers(size);
    printVector(vec);

    vector<int>answer(size);
    vector<int>cumulative;

    cout<<"**********Radix Sort**********\n";
    radixSort(vec,answer);
    printVector(answer);
    
    return 0;
}

File: 04_Non-ComparisonSorts/Radix Sort/RadixSort.py
import random


def initializeArrayWithRandomNumbers(n):
    listA = [0]
    for i in range(1, n):
        listA.append(random.randint(0, n))
    return listA


def countingSort(array, exp):
    n = len(array)
    outList = [0] * n
    cumFreq = [0] * 10

    for i in range(n):
        place = array[i] // exp
        cumFreq[place % 10] += 1

    for i in range(1, 10):
        cumFreq[i] += cumFreq[i - 1]

    i = n - 1
    while i >= 0:
        place = array[i] // exp
        outList[cumFreq[place % 10] - 1] = array[i]
        cumFreq[place % 10] -= 1
        i -= 1
    return outList


def radixSort(arr):
    m = max(arr)
    exp = 1
    while (m // exp) > 0:
        outList = countingSort(arr, exp)
        exp *= 10
    return outList


size = 10
print("Unsorted Array")
myList = initializeArrayWithRandomNumbers(size)
print(myList)

print("**********Radix Sort**********")
myList = radixSort(myList)
print(myList)
